daemon/src/model.rs

```diff
@@ -692,17 +690,145 @@ impl str::FromStr for OpeningFee {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         let amount_sat: u64 = s.parse()?;
-        Ok(OpeningFee { amount_sat })
+        Ok(OpeningFee {
+            fee: Amount::from_sat(amount_sat),
+        })
     }
 }
 
+impl_sqlx_type_display_from_str!(OpeningFee);
+
 impl Default for OpeningFee {
     fn default() -> Self {
-        Self::new(Amount::ZERO)
+        Self { fee: Amount::ZERO }
     }
 }
 
-impl_sqlx_type_display_from_str!(OpeningFee);
+impl From<OpeningFee> for Amount {
+    fn from(opening_fee: OpeningFee) -> Self {
+        opening_fee.fee
+    }
+}
+
+/// Funding fee as defined by position and rate
+///
+/// Position and rate define if the funding rate is to be added or subtracted when summing up the
+/// fees.
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct FundingFee {
+    #[serde(with = "bdk::bitcoin::util::amount::serde::as_sat")]
+    fee: Amount,
+    rate: FundingRate,
+}
+
+impl FundingFee {
+    pub fn new(fee: Amount, rate: FundingRate) -> Self {
+        Self { fee, rate }
+    }
+
+    pub fn into_signed_amount(self, position: Position) -> SignedAmount {
+        let fee: i64 = self.fee.as_sat().try_into().expect("not to overflow");
+
+        let signed = if (position == Position::Long && self.rate.0.is_sign_positive())
+            || (position == Position::Short && self.rate.0.is_sign_negative())
+        {
+            fee
+        } else {
+            -fee
+        };
+
+        SignedAmount::from_sat(signed)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum FeeFlow {
+    LongPaysShort,
+    ShortPaysLong,
+    Nein,
+}
+
+/// Our own accumulated fees
+///
+/// The counterparty fee is the inverse of this.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct FeeAccount {
+    balance: SignedAmount,
+    position: Position,
+    role: Role,
+}
+
+impl FeeAccount {
+    pub fn new(position: Position, role: Role) -> Self {
+        Self {
+            position,
+            role,
+            balance: SignedAmount::ZERO,
+        }
+    }
+
+    pub fn settle(&self) -> (Amount, FeeFlow) {
+        let flow = if self.balance == SignedAmount::ZERO {
+            FeeFlow::Nein
+        } else if (self.position == Position::Long && self.balance.is_positive())
```

Comment:
something seems not quite right if you need to expose the similar conditions in multiple places - can you use the same one from FundingFee or have a common abstraction? This seems like an implementation detail that shouldn't necessary leak to here.

Reply:
I would be in favor of moving all the decision logic into `FeeAccount` and have `OpeningFee` and `FundingFee` be relatively simple new types.

Reply:
yeah, the fact that we still need the logic in two places is a bit odd. This goes towards what @bonomat pointed out. We could actually already keep the `FeeFlow` internally in `FeeAccount` and only use settle. When adding we could still use a signed amount, but internally we could store it as absolute.

Reply:
How is storing it as absolute better?

A fee account is a subjective view (in contains a role and a position). Storing the signed amount is consistent with that. `settle` provides a different, objective view onto this data. Storing already that view would be weird.